### ISSUE TITLE
Rule proposal: avoid implicit truth checks for `number | undefined`

### ISSUE THREAD

**Issue by fregante** [typescript-eslint/typescript-eslint#8476]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

With:

```ts
type User {
  id?: number;
}
```

This will fail if the user has an `id` set to `0`

```js
if (!user.id) {
  console.log('Unregistered user')
}
```

Since we're checking for `undefined`, then it should be `=== undefined` instead, to avoid following into this trap

### Fail Cases

```TypeScript
function isRegistered(userId: number | undefined) {
  return Boolean(userId);
}
```

```ts
function isRegistered(userId: number | undefined) {
  if (userId) {
    return true;
  }

  return false;
}
```

```ts
function isRegistered(userId: number | undefined) {
  return userId ? true : false
}
```

### Pass Cases

```TypeScript
function isRegistered(userId: number | undefined) {
  return userId !== undefined
}
```

```ts
function isRegistered(userId: number | undefined) {
  if (typeof userId === 'number') {
    return true;
  }

  return false;
}
```

```ts
function isRegistered(userId: number | undefined) {
  return userId >= 0 ? true : false
}
```

### Additional Info

This rule is similar to https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/explicit-length-check.md, except that in this case it solves a bug, rather than being exclusively a readability issue.

**Comment by bradzacher** [comment]
You probably want [`strict-boolean-expressions`](https://typescript-eslint.io/rules/strict-boolean-expressions/). [playground](https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=HYQwtgpgzgDiDGEAEAzEBLANkg3gKCVQFdh4AXdAe2CXSgCUIBzOsiAJwgBMBGACiJQOASS4AuJMCJgARhyQAfJCS4QU6YNwCUuAoSScyRdjQBClSpgghgAoe1FaA3HoC%2BelCXJUadRiyg2Ti4AJjsRcUlpOXZFZWBVdU0uHXx9WhQkcIcU3XTCQ2MaMnYiCBd09z0CiCMTVBBMIQqkd0JPUgpqWgZmVg5uAGZs0QkpWXklFTUNbTz9QvrBCKQAfiQSsqQJNCaINzwq0EhYBGQ4KCh5ju9uvz7Agd4RyPGYuOmkubSF2qLleyiJAAQgAvKD4olZlwDu0vF1fL0AkFuGFljkxtFJpCZslUtUMlkyABPGAQSiZdFA8EQgDkbw4tPx%2BQMf3qm3KBKq6UWNF2zVhxE6Ph6-n6wWGVNeWNiUwSuO%2BBN5AJWAD4IQAGNYbUrIHaNIQHdxAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugIwHs%2BSAIZN6AD2LQUySnyaoMAbXDYciaND7RIAGhWrQ%2B7AF8jAXRWnTQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

The only case the rule doesn't handle is `Boolean()`.
This is intentional however because (a) it's not used too much in modern TS and (b) the rule only considers "condition" locations (c) its intended to be a clear and explicit escape hatch that says "I am intentionally coercing the value to a boolean".

**Comment by fregante** [comment]
Indeed! I saw that rule but it somehow did not check it :(
